xarray/plot/plot.py

```diff
@@ -555,14 +563,14 @@ def _infer_interval_breaks(coord):
 
 
 @_plot2d
-def pcolormesh(x, y, z, ax, **kwargs):
+def pcolormesh(x, y, z, ax, infer_intervals=True, **kwargs):
     """
     Pseudocolor plot of 2d DataArray
 
     Wraps matplotlib.pyplot.pcolormesh
     """
 
-    if not hasattr(ax, 'projection'):
+    if infer_intervals:
```

Comment:
Setting `infer_intervals=False` is exactly what you want if you have irregularly spaced coordinates (see example here: https://github.com/pydata/xarray/issues/781#issuecomment-191595449).

I think something like the logic I showed above may be a good compromise.  I'm happy to transition to always specifying `infer_intervals=False` we should be more transparent about the change (hence the warning along with a None for a default value.  Would you be happier with this logic?

```Python
if (infer_intervals is None):
    if (hasattr(ax, 'projection'))
        infer_intervals = False
        warnings.warn("infer_intervals=None is deprecated, please specify either True or False",
                      DeprecationWarning)
    else:
        infer_intervals = True
if infer_intervals:
    ...
```